pytorch_pfn_extras/training/extensions/slack.py

```diff
@@ -0,0 +1,205 @@
+
+        if not _requests_available and webhook_url is not None:
+            warnings.warn(
+                '`requests` package is unavailable. '
+                'Slack will do nothing.')
+            return
+
+        if webhook_url is None:
+            self._client = client
+            if client is None:
+                if token is None:
+                    token = os.environ.get('SLACK_TOKEN', None)
+                if token is None:
+                    raise RuntimeError(
+                        '`token` is needed for communicating with slack')
+                self._client = slack_sdk.WebClient(token=token)
+        else:
+            if (token, client) != (None, None):
+                raise ValueError(
+                    '`webhook_url` can\'t be used with client and token')
+            if filenames_template is not None:
+                raise ValueError(
+                    '`webhook_url` can\'t be used to post files'
+                    'please install `slack_sdk`')
+            self._client = None
+        self._webhook_url = webhook_url
+        # values in current observation or log report to send to slack
+        self._text = text_template
+        self._start_text = start_text_template
+        self._end_text = end_text_template
+        if filenames_template is None:
+            filenames_template = []
+        self._filenames = filenames_template
+        self._context = context_object
+        self._channel_id = channel_id
+        self._use_threads = use_threads
+        self._ts = None
+        self._trigger = trigger_module.get_trigger(trigger)
+
+    def _send_message(
+        self,
+        manager: ExtensionsManagerProtocol,
+        text: Union[
+            str, Callable[[ExtensionsManagerProtocol, Any, dict], str]
+        ]
+    ) -> None:
+        observation = manager.observation
+        if callable(self._text):
+            text = self._text(manager, self._context, observation)
+        else:
+            text = self._text.format(manager, self._context, **observation)
+
+        if self._webhook_url and _requests_available:
+            payload = {'text': text}
+            res = requests.post(self._webhook_url, json.dumps(payload))
+            if res.status_code != 200:
+                raise Exception(res.status_code, res.text)
+            return
+        elif not _requests_available:
+            return
+
+        if not _slack_sdk_available:
+            return
+
+        assert self._client is not None
+
+        ts = None
+        if self._use_threads:
+            ts = self._ts
+        response = self._client.chat_postMessage(
+            channel=self._channel_id,
+            text=text,
+            thread_ts=ts,
+        )
+        assert response.get("ok")  # type: ignore[no-untyped-call]
+        if self._use_threads and ts is None:
+            ts = response.get("ts")  # type: ignore[no-untyped-call]
+            self._ts = ts
+        for filename in self._filenames:
```

Comment:
Can we upload files along with the message?

Reply:
AFAIK they are two separated APIs, there is an `attachments` field in postMessage but the use is discouraged:

https://api.slack.com/messaging/composing/layouts#when-to-use-attachments

Also it seems to allow only URLs to files already uploaded somewhere.
 
After a light google search almost everyone resort to `files_upload`
https://stackoverflow.com/questions/66017386/cant-attach-uploaded-file-to-message-using-slack-api-via-python